libafl_unicorn/src/emu.rs

```diff
@@ -0,0 +1,308 @@
+                        begin = regions[i].begin;
+                    }
+                    if end > regions[i].end {
+                        end = regions[i].end;
+                    }
+
+                    let bytes = emu
+                        .mem_read_as_vec(begin, (end - begin) as usize)
+                        .expect("Could not get program code");
+
+                    let mut decoder = Decoder::with_ip(64, &bytes, begin, DecoderOptions::NONE);
+
+                    let mut formatter = NasmFormatter::new();
+                    formatter.options_mut().set_digit_separator("`");
+                    formatter.options_mut().set_first_operand_char_index(10);
+
+                    let mut instruction = Instruction::default();
+                    let mut output = String::new();
+
+                    while decoder.can_decode() {
+                        decoder.decode_out(&mut instruction);
+
+                        // Format the instruction ("disassemble" it)
+                        output.clear();
+                        formatter.format(&instruction, &mut output);
+
+                        let diff = instruction.ip() as i64 - pc as i64;
+                        print!("{:02}\t{:016X} ", diff, instruction.ip());
+                        let start_index = (instruction.ip() - begin) as usize;
+                        let instr_bytes = &bytes[start_index..start_index + instruction.len()];
+                        for b in instr_bytes.iter() {
+                            print!("{:02X}", b);
+                        }
+                        if instr_bytes.len() < HEXBYTES_COLUMN_BYTE_LENGTH {
+                            for _ in 0..HEXBYTES_COLUMN_BYTE_LENGTH - instr_bytes.len() {
+                                print!("  ");
+                            }
+                        }
+                        println!(" {}", output);
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn callback(
+    _unicorn: &mut unicorn_engine::Unicorn<()>,
+    mem: MemType,
+    address: u64,
+    size: usize,
+    value: i64,
+) -> bool {
+    if DEBUG {
+        match mem {
+            MemType::WRITE => println!(
+                "Memory is being WRITTEN at adress: {:X} size: {} value: {}",
+                address, size, value
+            ),
+            MemType::READ => println!(
+                "Memory is being READ at adress: {:X} size: {}",
+                address, size
+            ),
+
+            _ => println!(
+                "Memory access type: {:?} adress: {:X} size: {} value: {}",
+                mem, address, size, value
+            ),
+        }
+    }
+
+    return true;
+}
+
+pub fn prog(emu: &mut unicorn_engine::Unicorn<'static, ()>, arm_code_len: u64) {
+    let result = emu.emu_start(
+        match emu.get_arch() {
+            Arch::ARM64 => CODE_ADDRESS + 0x40, // Position of main: 0x40 TODO: see if possible to get the main position from header file. Seems weird doing so
```

Comment:
TODO: delete this overide
This was added as in the text program the section 0x0 - 0x3F contains invalid code